### PR TITLE
Introduce ReferenceFieldBase

### PR DESCRIPTION
**Pull request by djhi**
## Problem

While working on the alternative UI libraries demos, we had to re-implement all the reference fields and inputs logic.

## Solution

Add core headless versions for these components
This PR introduces only the `ReferenceFieldBase`.

## Tasks

- [x] Implementation
- [x] Tests

### CODE REVIEW

**Review by erwanMarmelab — APPROVED** [review]

File: packages/ra-core/src/controller/field/useReferenceFieldController.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import { useResourceDefinition } from '../../core';
+
+export const useReferenceFieldController = <
+    RecordType extends Record<string, any> = Record<string, any>,
+    ReferenceRecordType extends RaRecord = RaRecord
+>(
+    options: UseReferenceFieldControllerOptions<ReferenceRecordType>
+): UseReferenceFieldControllerResult<ReferenceRecordType> => {
+    const { source, link = 'edit', reference, queryOptions } = options;
+    const record = useRecordContext<RecordType>(options);
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/field/useReferenceFieldController.ts:17]: throw if reference is empty

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: packages/ra-ui-materialui/src/field/ReferenceField.tsx [hunk 2]
```diff
@@ -63,26 +57,13 @@ export const ReferenceField = <
...
-    const id = useFieldValue(props);
-
-    const translate = useTranslate();
-
-    return id == null ? (
-        emptyText ? (
-            <Typography component="span" variant="body2">
-                {emptyText && translate(emptyText, { _: emptyText })}
-            </Typography>
-        ) : null
```
[the remark this answers is not in the document]
Comment (djhi) [reply packages/ra-ui-materialui/src/field/ReferenceField.tsx:77]: Storybook disagrees with you. Having no `id` will disable the query in `useReferenceFieldController` so no `referenceRecord` will be available at L146 in this file thus showing the empty message

[the remark this answers is not in the document]
Comment (djhi) [reply packages/ra-ui-materialui/src/field/ReferenceField.tsx:77]: There was an error in the `useReferenceFieldController` though. Both the tests and the storybook show that all empty cases are handled

[the remark this answers is not in the document]
Comment (fzaninotto) [reply packages/ra-ui-materialui/src/field/ReferenceField.tsx:77]: Except in the previous code, the empty text was displayed in a `Typography`. That's no longer the case.

**Review by fzaninotto — APPROVED** [review]

**Pull request merged**